### PR TITLE
Add limits for number of transactions by sender

### PR DESCRIPTION
**Pull request by andreibancioiu**
A sender can have no more than `k` transactions in the pool. The transactions can not exceed `s` bytes in total. These limits are checked and enforced at transaction insertion (addition) time.

```
k: SizePerSender = 1000
s: SizeInBytesPerSender = 614400
```

Furthermore, removed phase 1 of eviction ("evict high nonce transactions") - redundant. The previous phase 2 ("evict senders with bad score") is now phase 1.

Lastly, fixed the fee scale when computing the score of the senders - because of the new `minGasPrice` (the change of 3 orders of magnitude affected the score formula).

**Note that this PR includes https://github.com/ElrondNetwork/elrond-go/pull/1725, so that one should enter first in development - thus please skip testnet-related changes when reviewing this PR.**

### CODE REVIEW

File: storage/txcache/wrappedTransaction.go [hunk 1]
```diff
@@ -9,3 +13,7 @@ type WrappedTransaction struct {
 	SenderShardID   uint32
 	ReceiverShardID uint32
 }
+
+func (wrappedTx *WrappedTransaction) sameAs(another *WrappedTransaction) bool {
```
Comment (iulianpascalau) [inline storage/txcache/wrappedTransaction.go:17]: :+1:

File: process/smartContract/process.go [hunk 2]
```diff
@@ -540,7 +540,8 @@ func (sc *scProcessor) DeploySmartContract(
 
 func (sc *scProcessor) printScDeployed(vmOutput *vmcommon.VMOutput, tx data.TransactionHandler) {
 	scGenerated := make([]string, 0, len(vmOutput.OutputAccounts))
-	for addr := range vmOutput.OutputAccounts {
+	for key := range vmOutput.OutputAccounts {
+		addr := vmOutput.OutputAccounts[key].Address
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/smartContract/process.go:544]: Was fixed in @camilbancioiu's PR (merged here).

File: scripts/testnet/variables.sh [hunk 3]
```diff
@@ -34,9 +34,10 @@ export NODE_NICENESS=10
 
 # Delays after running executables.
 export SEEDNODE_DELAY=5
-export NODE_DELAY=10
+export GENESIS_DELAY=30
+export NODE_DELAY=$GENESIS_DELAY
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply scripts/testnet/variables.sh:38]: Was reverted in @camilbancioiu's PR (merged here).

File: storage/txcache/config.go [hunk 4]
```diff
@@ -1,14 +1,54 @@
...
-	NumTxsToEvictFromASender   uint32
-	MinGasPriceMicroErd        uint32
+	MinGasPriceNanoErd         uint32
+}
+
+func (config *CacheConfig) verify() error {
+	if len(config.Name) == 0 {
+		return fmt.Errorf("%w: config.Name is invalid", errInvalidCacheConfig)
+	}
+
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/txcache/config.go:22]: Fixed.

File: storage/txcache/txCache_test.go [hunk 5]
```diff
@@ -13,8 +13,72 @@ import (
...
+	requireErrorOnNewTxCache(t, badConfig, "config.CountThreshold")
+
+	badConfig = withEvictionConfig
+	badConfig.NumSendersToEvictInOneStep = 0
+	requireErrorOnNewTxCache(t, badConfig, "config.NumSendersToEvictInOneStep")
+}
+
+func requireErrorOnNewTxCache(t *testing.T, config CacheConfig, errMessage string) {
+	cache, err := NewTxCache(config)
+	require.Contains(t, err.Error(), errMessage)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/txcache/txCache_test.go:76]: Fixed - if I understood well :)

File: dataRetriever/txpool/argShardedTxPool.go [hunk 6]
```diff
@@ -21,9 +21,15 @@ func (args *ArgShardedTxPool) verify() error {
 	if config.SizeInBytes < dataRetriever.TxPoolMinSizeInBytes {
 		return fmt.Errorf("%w: config.SizeInBytes is less than [dataRetriever.TxPoolMinSizeInBytes]", dataRetriever.ErrCacheConfigInvalidSizeInBytes)
 	}
+	if config.SizeInBytesPerSender < dataRetriever.TxPoolMinSizeInBytes {
+		return fmt.Errorf("%w: config.SizeInBytesPerSender is less than [dataRetriever.TxPoolMinSizeInBytes]", dataRetriever.ErrCacheConfigInvalidSizeInBytes)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply dataRetriever/txpool/argShardedTxPool.go:25]: `TxPoolMinSizeInBytes` is (was) ~ 40kb, and added mostly to ensure that there are no zeros (just positives) in the configuration. This was an arbitrary number to account for the size in bytes of one transaction, indeed (like 1 was used for counts / num shards and so on).

Fixed to remove ambiguity, and now we use a simple condition for "strictly positive".

File: storage/txcache/disabledCache.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+// DisabledCache represents a disabled cache
+type DisabledCache struct {
+}
+
+// NewDisabledCache creates a new disabled cache
+func NewDisabledCache() *DisabledCache {
+	return &DisabledCache{}
+}
+
+// AddTx -
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/txcache/disabledCache.go:18]: Fixed comments. Indeed, logging of errors not mandatory (we already have `log.Error("shardedTxPool.createTxCache()", "err", err)` for errors when creating, say, a cache with invalid configuration - which would have resulted after all in creating a disabled cache instead). Therefore, fixed, removed logs.

File: storage/txcache/txListForSender.go [hunk 8]
```diff
@@ -61,32 +62,81 @@ func (listForSender *txListForSender) AddTx(tx *WrappedTransaction) {
...
+		if incomingTx.sameAs(tx) {
+			// The incoming transaction will be discarded
+			return nil, errTxDuplicated
+		}
 
 		if nonce == incomingNonce && gasPrice > incomingGasPrice {
-			return element
+			// The incoming transaction will be placed right after the existing one, which has same nonce but higher price.
+			return element, nil
 		}
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/txcache/txListForSender.go:130]: Code in `findInsertionPlace` isn't currently very self-explanatory, but if the `I've found a place` event does not happen in the iteration where the condition `iteratedTxNonce == incomingNonce && iteratedTxGasPrice <= incomingGasPrice` holds, the flow goes on and then stops at `iteratedTxNonce < incomingNonce` which is the most brutal / catch all condition of the iteration - and triggers the `I've found a place` event (and thus the SORT BY nonce ASC THEN BY GAS DESC holds).

Fixed ambiguity by adding an extra test and some comments.

**Pull request merged**